### PR TITLE
Option to add a debug header to see which route/action/director/backe…

### PR DESCRIPTION
**Pull request by Pols95**
Now, in the configuration can be setup **mapper.debug=true|false** (default false) and **mapper.debug.name=header-name** (default Routing-Path) to add a custom header **X-header-name: id_selected_route;id_selected_action;id_selected_director;id_selected_backend** to HttpResponses for actions of type *cache* and *proxy*

### CODE REVIEW

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 1]
```diff
@@ -367,37 +373,38 @@ public MapResult map(HttpRequest request, String userId, String sessionId, Backe
...
                     }
+
+                    BackendConfiguration backend = this.backends.get(backendId);
+                    if (backend != null && backendHealthManager.isAvailable(backendId)) {
+                        if (this.debuggingHeaderEnabled) {
+                            String routingPath = "Route-id: " + route.getId() + "; ";
+                            routingPath += "Action-id: " + action.getId() + "; ";
+                            routingPath += "Director-id: " + action.getDirector() + "; ";
+                            routingPath += "Backend-id: " + backendId;
+                            action.addCustomHeader(new CustomHeader(DEBUGGING_HEADER_ROUTING_PATH, routingPath, HeaderMode.HEADER_MODE_ADD));
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: this is not good, "Actions" are immutable objects, you cannot add a "custom header"

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 2]
```diff
@@ -99,6 +103,8 @@ public void configure(ConfigurationStore properties) throws ConfigurationNotVali
         LOG.info("configured mapper.forcedirector.parameter=" + forceDirectorParameter);
         this.forceBackendParameter = properties.getProperty("mapper.forcebackend.parameter", forceBackendParameter);
         LOG.info("configured mapper.forcebackend.parameter=" + forceBackendParameter);
+        this.debuggingHeaderEnabled = Boolean.parseBoolean(properties.getProperty("mapper.debuggingheader.enable", "false"));
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: what about "mapper.debug" ?

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 3]
```diff
@@ -75,6 +76,9 @@
     private static final String ACME_CHALLENGE_URI_PATTERN = "/\\.well-known/acme-challenge/";
     private DynamicCertificatesManager dynamicCertificateManger;
 
+    public static final String DEBUGGING_HEADER_ROUTING_PATH = "Routing-Path";
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: better to make it configurable
the debug MUST be a name which starts with "X-", which means it is a custom header

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 4]
```diff
@@ -367,37 +373,38 @@ public MapResult map(HttpRequest request, String userId, String sessionId, Backe
...
                             break;
                         default:
                             return MapResult.NOT_FOUND(route.getId());
                     }
+
+                    BackendConfiguration backend = this.backends.get(backendId);
+                    if (backend != null && backendHealthManager.isAvailable(backendId)) {
+                        if (this.debuggingHeaderEnabled) {
+                            String routingPath = "Route-id: " + route.getId() + "; ";
+                            routingPath += "Action-id: " + action.getId() + "; ";
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: this is too verbose, just "id;id;id;id" is enough

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java (hunk 1 above)
Comment (Pols95) [reply carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: Do you think it's ok cloning the object ('cause i think it's good adding all additional headers at this higher level as CustomHeader)? Or is it better to add it in the request handler (lower level)?

Comment (eolivelli) [reply carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: Clone = allocate...
If we can save allocations on the hot path it is better.
So I prefer not to clone

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 5]
```diff
@@ -99,6 +104,11 @@ public void configure(ConfigurationStore properties) throws ConfigurationNotVali
         LOG.info("configured mapper.forcedirector.parameter=" + forceDirectorParameter);
         this.forceBackendParameter = properties.getProperty("mapper.forcebackend.parameter", forceBackendParameter);
         LOG.info("configured mapper.forcebackend.parameter=" + forceBackendParameter);
+        // To add custom debugging header for request choosen mapping-path
+        this.debuggingHeaderEnabled = Boolean.parseBoolean(properties.getProperty("mapper.debug", "false"));
+        LOG.info("configured mapper.debug=" + debuggingHeaderEnabled);
+        this.debuggingHeaderName = "X-" + properties.getProperty("mapper.debug.name", DEBUGGING_HEADER_DEFAULT_NAME);
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: this way you are forcing that prefix even on the user-configured value.
It is better to have a default of "X-Proxy-Path" and let the user override the value

File: carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java [hunk 6]
```diff
@@ -367,37 +377,39 @@ public MapResult map(HttpRequest request, String userId, String sessionId, Backe
...
+                            selectedAction = MapResult.Action.CACHE;
                             break;
                         default:
                             return MapResult.NOT_FOUND(route.getId());
                     }
+
+                    BackendConfiguration backend = this.backends.get(backendId);
+                    if (backend != null && backendHealthManager.isAvailable(backendId)) {
+                        List<CustomHeader> customHeaders = new ArrayList(action.getCustomHeaders());
+                        if (this.debuggingHeaderEnabled) {
```
Comment (eolivelli) [inline carapace-server/src/main/java/org/carapaceproxy/server/mapper/StandardEndpointMapper.java]: if this debug is not enabled you can skip the clone of the arraylist.

please also ensure that action.getCustomHeaders() returns an unmodifiableList

**Pull request merged**